File: gfapy/placeholder.py
import gfapy

class Placeholder:
  """
  A placeholder is used in mandatory fields when a value is not specified.

  Its string representation is an asterix ``*``.
  """

  def __str__(self):
    return "*"

  def __repr__(self):
    return "gfapy.Placeholder()"

  def __bool__(self):
    return False

  def complement(self):
    return self

  def is_empty(self):
    """
    A placeholder is always empty.

    Returns
    -------
    True : bool
      Returns always **True**
    """
    return True

  def validate(self, *args, **keyargs):
    """
    A placeholder is always valid.
    """
    return None

  def rc(self):
    """
    For compatibility with String#rc (gfapy.sequence module).

    Returns
    -------
    self : Placeholder
    """
    return self

  def __len__(self):
    """
    Length/size of a placeholder is always 0.

    Returns
    -------
    0 : int
      Returns always 0
    """
    return 0

  def __getitem__(self, key):
    """
    Any cut of the placeholder returns the placeholder itself.

    Parameters
    ----------
    key
      Ignored

    Returns
    -------
    self : Placeholder
    """
    return self

  def __add__(self, other):
    """
    Adding the placeholder to anything returns the placeholder itself.

    Parameters
    ----------
    other
      Ignored

    Returns
    -------
    self : Placeholder
    """
    return self

  def __eq__(self, other):
    return gfapy.is_placeholder(other)

def is_placeholder(object):
  if object is Placeholder:
    return True
  elif object is None:
    return True
  elif object == "*":
    return True
  elif isinstance(object, list) and len(object) == 0:
    return True
  else:
    return False

File: lib/rgfa/placeholder.rb
# A placeholder is used in mandatory fields when a value is not specified.
# Its string representation is an asterisk (+*+).
class RGFA::Placeholder
  # @return [String] string representation (+*+)
  def to_s
    "*"
  end

  # A placeholder is always empty
  # return [true]
  def empty?
    true
  end

  # A placeholder is always valid
  # return [void]
  def validate
  end

  # For compatibility with String#rc (RGFA::Sequence module)
  # @return [self]
  def rc
    self
  end

  # Length/size of a placeholder is always 0
  # @return [self]
  def length
    0
  end

  # @return [true]
  def placeholder?
    true
  end

  alias_method :size, :length

  # Any cut of the placeholder returns the placeholder itself
  # @param anything [Object] ignored
  # @return [self]
  def [](*anything)
    self
  end

  # Adding the placeholder to anything returns the placeholder itself
  # @param anything [Object] ignored
  # @return [self]
  def +(*anything)
    self
  end

  def ==(other)
    other.placeholder?
  end

  alias :eql? :==
  alias :=== :==

end

class String
  def placeholder?
    self == "*"
  end
end

class Symbol
  def placeholder?
    self == :"*"
  end
end

class Array
  def placeholder?
    empty?
  end
end

class Numeric
  def placeholder?
    false
  end
end
